File: ms-viagens/src/main/java/br/edu/ifsp/carona/viagens/entity/Viagem.java
package br.edu.ifsp.carona.viagens.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "viagens")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Viagem {

    public enum Status {
        ABERTA, EM_ANDAMENTO, CONCLUIDA, CANCELADA
    }

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private Long motoristaId;

    @Column(nullable = false)
    private String origem;

    @Column(nullable = false)
    private String destino;

    @Column(nullable = false)
    private LocalDateTime dataHoraSaida;

    @Column(nullable = false)
    private Integer vagas;

    @Builder.Default
    private Integer vagasOcupadas = 0;

    @Column(nullable = false)
    private BigDecimal valorContribuicao;

    @Enumerated(EnumType.STRING)
    @Builder.Default
    private Status status = Status.ABERTA;

    @CreationTimestamp
    private LocalDateTime criadoEm;

    @UpdateTimestamp
    private LocalDateTime atualizadoEm;
}

File: mobile/lib/models/viagem.dart
class Viagem {
  final int? id;
  final int motoristaId;
  final String origem;
  final String destino;
  final String dataHoraSaida;
  final int vagas;
  final int vagasOcupadas;
  final double valorContribuicao;
  final String status;
  final String? criadoEm;

  Viagem({
    this.id,
    required this.motoristaId,
    required this.origem,
    required this.destino,
    required this.dataHoraSaida,
    required this.vagas,
    required this.vagasOcupadas,
    required this.valorContribuicao,
    required this.status,
    this.criadoEm,
  });

  factory Viagem.fromJson(Map<String, dynamic> json) {
    return Viagem(
      id: json['id'] as int?,
      motoristaId: json['motoristaId'] as int? ?? 0,
      origem: json['origem'] as String? ?? '',
      destino: json['destino'] as String? ?? '',
      dataHoraSaida: json['dataHoraSaida'] as String? ?? '',
      vagas: json['vagas'] as int? ?? 0,
      vagasOcupadas: json['vagasOcupadas'] as int? ?? 0,
      valorContribuicao: (json['valorContribuicao'] as num?)?.toDouble() ?? 0.0,
      status: json['status'] as String? ?? 'ABERTA',
      criadoEm: json['criadoEm'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'motoristaId': motoristaId,
      'origem': origem,
      'destino': destino,
      'dataHoraSaida': dataHoraSaida,
      'vagas': vagas,
      'vagasOcupadas': vagasOcupadas,
      'valorContribuicao': valorContribuicao,
      'status': status,
      'criadoEm': criadoEm,
    };
  }
}
